src/site/pages/digital/src/utils/ExpressionParser/TreeToCircuit.ts

```diff
@@ -0,0 +1,130 @@
+import {InputTree} from "./Constants/DataStructures";
+
+import {CreateCircuit, DigitalCircuit}    from "digital/public";
+import {V}                                from "Vector";
+import {Component, Port}                  from "core/public";
+import {Err, ErrE, Ok, Result,ResultUtil} from "core/utils/Result";
+
+
+/**
+ * Used to get the string for the Create funciton from the operation.
+ */
+export const TypeToGate = {
+    "&": "ANDGate",
+    "!": "NOTGate",
+    "|": "ORGate",
+    "^": "XORGate",
+} as const;
+
+/**
+ * Used to get the string for the Create funciton from the negated operation.
+ */
+export const NegatedTypeToGate = {
+    "&": "NANDGate",
+    "|": "NORGate",
+    "^": "XNORGate",
+} as const;
+
+/**
+ * Connects the first available output port on `prevComp` to the `newNode` on `newComp`.
+ *
+ * @param prevComp The component to connect the output of.
+ * @param newNode  First available `inputs` port on `newComp`.
+ * @param newComp  The component to connect the input of, only used for error message and return value.
+ * @returns        In the case of success, a `Result` wrapping `newComp`. An error in the following cases:
+ *                 - No output port found on `prevComp`.
+ *                 - Wire failed to create connecting `prevComp` to `newNode`.
```

Comment:
Can these failure cases ever occur?  It's preferable to throw exceptions and use `.unwrap()` in these cases if the `Err` path can only be triggered by internal bugs (this comment applies generally to this file).

Reply:
Yeah I got a bit overzealous with Result, I don't think it actually applies to this step. None of the error paths here should be triggered by user input, only by something going wrong internally.